### DISCUSSION TITLE
connect event not called after a delay celery task

### DISCUSSION THREAD

**Discussion by cris-m** [miguelgrinberg/Flask-SocketIO#1888] [Q&A] [answered]
Here is the `create_app` method of Flask app. I have a separate file where all the app extension are initialized
```python
# extensions.py file
from flask_socketio import SocketIO

socketio = SocketIO(manage_session=False,
                    message_queue='redis://redis:6379/0',
                    async_mode="threading",
                    cors_allowed_origins='*',
                    logger=True,
                    engineio_logger=True)

```
```python
# app.py file
from flask import Flask
from app.exetensions import socketio

def create_app():
     app = Flask(__name__, instance_relative_config=True)
     socketio.init_app(app,  
          message_queue='redis://redis:6379/0', 
          async_mode='threading',  
          logger=True,
          engineio_logger=True)
   
      @socketio.event
      def connect():
          app.logger.info('[*] Connect event')
          socketio.emit('after_connect', {'data': 'some data'})

      if __name__ == '__main__':
          socketio.run(app)
     
      return app
```
The app is execute from `docker-compose` using the following command:
```
gunicorn -c python:app.config.gunicorn -k geventwebsocket.gunicorn.workers.GeventWebSocketWorker -w 1 app.app:create_app()
```
Here is some log after the celery task
```
emitting event "file" to bxj4rqRx5wZJRI0BAAAD [/]
[2022-10-04 11:37:22,290: INFO/ForkPoolWorker-2] emitting event "file" to bxj4rqRx5wZJRI0BAAAD [/]
pubsub message: emit
emitting event "process" to bxj4rqRx5wZJRI0BAAAD [/]
[2022-10-04 11:37:22,295: INFO/ForkPoolWorker-2] emitting event "process" to bxj4rqRx5wZJRI0BAAAD [/]
pubsub message: emit
[2022-10-04 11:37:22,303: INFO/ForkPoolWorker-2] Task app.models.files.tasks.process_images[457d3f53-4ba6-410f-a72b-42080a64984d] succeeded in 14.010606162999466s: None
ysBWzhtnu9kcr8KAAAI: Sending packet PING data None
ysBWzhtnu9kcr8KAAAI: Received packet PONG data
ysBWzhtnu9kcr8KAAAI: Sending packet PING data None
ysBWzhtnu9kcr8KAAAI: Received packet PONG data
ysBWzhtnu9kcr8KAAAI: Sending packet PING data None
ysBWzhtnu9kcr8KAAAI: Received packet PONG data
```
When the app  execute a celery task for the fist time, everything goes well but after the celery task, if the browser is refresh, `create` event does not get call but the app refresh the client `SID` and I see the following log:
```
The WebSocket transport is not available, you must install a WebSocket server that is compatible with your async mode to enable it. See the documentation for details.
```
. The client listening for `after_connect` event but after the celery event, it does not receive the `emit` message from server.

**Comment by miguelgrinberg** [comment, marked as answer]
You are using the threading async mode. Why are you running gunicorn with the gevent worker?

**Reply by cris-m** [reply]
Without using `gevent` I am getting `BAD REQUEST` from `flask_socketio`. with log like the following:
```
HjVgCME5OpvmcFovAAAt: Sending packet OPEN data {'sid': 'HjVgCME5OpvmcFovAAAt', 'upgrades': [], 'pingTimeout': 20000, 'pingInterval': 25000}
```
I don't understand why the `upgrades` is an empty list

**Reply by miguelgrinberg** [reply]
Upgrades is empty because you do not have a websocket server installed. This is the error you mentioned above. If you use the threading mode, you have to install simple-websocket to have the ws protocol available. This is all covered in the deployment section of the docs, by the way.

**Reply by cris-m** [reply]
I have removed `gevent ` worker but I am getting the same behavior. After the celery task. The `connect` event will not be called even the browser is refresh multiple time. This is the log which I collect after refreshing the browser.
```
99854SFYKj8aYfGqAAAA: Sending packet PING data None
99854SFYKj8aYfGqAAAA: Received packet PONG data
99854SFYKj8aYfGqAAAA: Unexpected error "", closing connection
172.18.0.6 - - [04/Oct/2022:16:04:27 +0000] 'GET /socket.io/?EIO=4&transport=websocket&sid=99854SFYKj8aYfGqAAAA HTTP/1.1' 200 0 '-' 'Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/106.0.0.0 Mobile Safari/537.36' in 77472415µs
S2nzgdUcPyZwW-cxAAAC: Sending packet OPEN data {'sid': 'S2nzgdUcPyZwW-cxAAAC', 'upgrades': ['websocket'], 'pingTimeout': 20000, 'pingInterval': 25000}
172.18.0.6 - - [04/Oct/2022:16:04:27 +0000] 'GET /socket.io/?EIO=4&transport=polling&t=OEZlSh3 HTTP/1.1' 200 97 'http://localhost/' 'Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/106.0.0.0 Mobile Safari/537.36' in 2770µs
172.18.0.6 - - [04/Oct/2022:16:04:28 +0000] 'GET /api/v1/user HTTP/1.1' 200 276 'http://localhost/' 'Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/106.0.0.0 Mobile Safari/537.36' in 14568µs
S2nzgdUcPyZwW-cxAAAC: Received request to upgrade to websocket
172.18.0.6 - - [04/Oct/2022:16:04:28 +0000] 'GET /socket.io/?EIO=4&transport=polling&t=OEZlSiH&sid=S2nzgdUcPyZwW-cxAAAC HTTP/1.1' 200 1 'http://localhost/' 'Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/106.0.0.0 Mobile Safari/537.36' in 788µs
172.18.0.6 - - [04/Oct/2022:16:04:28 +0000] 'GET /socket.io/?EIO=4&transport=polling&t=OEZlSjD&sid=S2nzgdUcPyZwW-cxAAAC HTTP/1.1' 200 1 'http://localhost/' 'Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/106.0.0.0 Mobile Safari/537.36' in 644µs
S2nzgdUcPyZwW-cxAAAC: Received packet MESSAGE data 0
S2nzgdUcPyZwW-cxAAAC: Sending packet MESSAGE data 0{"sid":"Mnq9tCXG5yMsbi7TAAAD"}
172.18.0.6 - - [04/Oct/2022:16:04:28 +0000] 'POST /socket.io/?EIO=4&transport=polling&t=OEZlSiG&sid=S2nzgdUcPyZwW-cxAAAC HTTP/1.1' 200 2 'http://localhost/' 'Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/106.0.0.0 Mobile Safari/537.36' in 2781µs
S2nzgdUcPyZwW-cxAAAC: Upgrade to websocket successful
S2nzgdUcPyZwW-cxAAAC: Sending packet PING data None
```
I can see there is a new SID but the `connect` event method from the server was not called.

**Reply by miguelgrinberg** [reply]
What do you expect a connection after a celery task runs? What does celery have to do with your client reconnecting? I'm sorry, but I'm very confused, I don't understand what you think is a problem here. Maybe you need to create a small and complete example application that demonstrates your problem, so that I can test it here.

**Reply by cris-m** [reply]
For every reconnect, I keep the client SID in redis on each connection (in `connect` event).  When the server need to send the client a private message, it gets the client SID in redis and send the private message. If the client reconnect and `connect` method does not execute, redis will have the old SID and the server private method will not reach to the client since the SID is different. What strange is the client reconnect but the connect event does not execute. 
```python
# app.py file 
from flask import Flask
from app.exetensions import socketio, redis

def create_app():
     app = Flask(__name__, instance_relative_config=True)
     socketio.init_app(app,  
          message_queue='redis://redis:6379/0', 
          async_mode='threading',  
          logger=True,
          engineio_logger=True)
   
      @socketio.event
      def connect():
          app.logger.info('[*] Connect event')
          user_id = current_user.id
          redis.set(user_id, request.sid)
          socketio.emit('after_connect', room=request.sid)

      if __name__ == '__main__':
          socketio.run(app)
     
      return app
```
And here is he  celery task example
```python
# tasks.py
from flask import current_user
from app.app import create_celery_app
from app.exetensions import socketio, redis

celery = create_celery_app()

@celery.task()
def example_task():
    user_id = current_user.id
    user_sid = redis.get(user_id).decode()
    # perform some task
    socket.emit('task_finish', {'data': 'task finish'}, room=user_sid)
```

Here is the log collected after the celery task
```
emitting event "task_finish" to Tk9tzlhQHxKPrtMQAAAF [/]
[2022-10-05 04:43:00,445: INFO/ForkPoolWorker-2] emitting event "file" to Tk9tzlhQHxKPrtMQAAAF [/]
pubsub message: emit
hdTalRU_lAWOhMCcAAAE: Sending packet MESSAGE data 2["task_finish",{"data": "task finish"}}]
[2022-10-05 04:43:00,456: INFO/ForkPoolWorker-2] Task app.tasks[b3fe5b6e-0e8d-43d9-bb11-00c21f3638ae] succeeded in 8.299068402999183s: None
hdTalRU_lAWOhMCcAAAE: Sending packet PING data None
hdTalRU_lAWOhMCcAAAE: Received packet PONG data
57rhrW-V5FTsqRT5AAAB: Client is gone, closing socket
57rhrW-V5FTsqRT5AAAB: Client is gone, closing socket
uCdbAHeqhNjFb-_UAAAA: Client is gone, closing socket
uCdbAHeqhNjFb-_UAAAA: Client is gone, closing socket
```

**Reply by miguelgrinberg** [reply]
You are not answering my question. Why does your client reconnect after running a Celery task? I don't understand why are we talking about Celery at all in this context.

**Reply by cris-m** [reply]
On each browser refresh the `connect` method execute and the `SID` change. That how the socketio works.  In my question I am talking about celery because everything works perfectly until the celery task execute. It seems like `gunicorn` enter a block state due `threadings` on 1 worker  and all the socket connection.

**Reply by miguelgrinberg** [reply]
I think if Gunicorn and or Celery cause blocking in your application, it is probably a bug in your own code. Could you please take any of the examples in this repository and expand it to demonstrate the issue? That is the only way I can test this here.

**Reply by cris-m** [reply]
I will work on an example on this repo. To check if I can reproduce the issue

**Reply by cris-m** [reply]
I added the `Flask-Socketio-Chat` app in docker to follow the structure I have on my project. I have reproduced the issues I am having by adding one link in `chat` route to execute a background celery task.  To see the issue flow the following step:
1. Run the app
2. Join the chat room. (Everything work fine). Messaging will work perfectly
3. Click on `Execute celery task` link (all the `joined`, `text` event will be send but the server will not execute the event function)

[flask_sockeio_test](https://github.com/cris-m/flask_sockeio_test) repo

**Reply by miguelgrinberg** [reply]
You are using an invalid gunicorn configuration for eventlet. See the documentation for how to run gunicorn when using eventlet.

**Reply by cris-m** [reply]
> You are using an invalid gunicorn configuration for eventlet. See the documentation for how to run gunicorn when using eventlet.

Here a simple configuration I am using.  I have set the `worker_class` to `geventwebsocket`
```python
import os
from distutils.util import strtobool

from gevent import monkey
monkey.patch_all()

bind = f'0.0.0.0:{os.getenv("WEB_API_PORT", "8000")}'
loglevel = 'debug'
accesslog = '-'
access_log_format = "%(h)s %(l)s %(u)s %(t)s '%(r)s' %(s)s %(b)s '%(f)s' '%(a)s' in %(D)sµs"  # noqa: E501
workers = 1
worker_class = "geventwebsocket.gunicorn.workers.GeventWebSocketWorker"
threads = 100
reload = bool(strtobool(os.getenv('WEB_RELOAD', 'false')))
```

**Reply by miguelgrinberg** [reply]
I'm sorry, but we are confusing things here. You gave me an example application to debug. I'm commenting on that application.

I've found several problems in this example that you provided, too many to really even try to make it work:

- Gunicorn is started in threading mode. The Flask app is running in threading mode (which is fine), but the Celery app is started in default mode, which would be eventlet, since you are installing eventlet in your application. This will never work.
- You are not connecting the Socket.IO instance to the message queue.
- You start the celery task by clicking a link. This breaks the chat connection that was established in the page.
- You are creating `socketio` and `celery_app` instances in multiple places. You have to have a single instance of each in the Flask app, and a single instance of each in the Celery app. You cannot cross import files from one to the other, because then multiple instances are created in each process.

I suggest you look at my Flack (https://github.com/miguelgrinberg/flack) example if you want to learn a clean way to create a Socket.IO project that uses Celery.

**Reply by cris-m** [reply]
> I'm sorry, but we are confusing things here. You gave me an example application to debug. I'm commenting on that application.
> 
> I've found several problems in this example that you provided, too many to really even try to make it work:
> 
> * Gunicorn is started in threading mode. The Flask app is running in threading mode (which is fine), but the Celery app is started in default mode, which would be eventlet, since you are installing eventlet in your application. This will never work.
> * You are not connecting the Socket.IO instance to the message queue.
> * You start the celery task by clicking a link. This breaks the chat connection that was established in the page.
> * You are creating `socketio` and `celery_app` instances in multiple places. You have to have a single instance of each in the Flask app, and a single instance of each in the Celery app. You cannot cross import files from one to the other, because then multiple instances are created in each process.
> 
> I suggest you look at my Flack (https://github.com/miguelgrinberg/flack) example if you want to learn a clean way to create a Socket.IO project that uses Celery.

Yes, Socket.IO instance was not connected to message queue. I have fixed it. Thanks for the `Flack` link. I will check it to make a clean app
